File: Osnovne skole/2024/Alinina abeceda/Alinina_abeceda.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

string alphabet;
vector<int> val(26);

bool compare(const string &a, const string &b) {
    if(a.length() > b.length()) return !compare(b, a);
    
    for(int i = 0; i < a.length(); i++) {
        if(a[i] == b[i]) continue;

        char AA = 'a';
        if(i == 0) AA = 'A';

        return val[a[i]-AA] < val[b[i]-AA];
    }
    
    if(a.length() == b.length()) return false;

    return true;
}

int main() {
    cin >> alphabet;
    for(int i = 0 ; i < alphabet.length(); i++) {
        char curr = alphabet[i];
        val[curr - 'a'] = i;
    }

    int n; cin >> n;
    vector<string> words(n);
    for(int i = 0; i < n; i++)
        cin >> words[i];
    
    sort(words.begin(), words.end(), compare);
            
    for(int i = 0 ; i < n; i++)
        cout << words[i] << endl;
    return 0;
}

File: Osnovne skole/2024/Alinina abeceda/Alinina_abeceda.py
def compare(a, b):
    if len(a) > len(b):
        return not compare(b, a)
    
    for i in range(len(a)):
        if a[i] == b[i]:
            continue
        
        AA = 'a'
        if i == 0:
            AA = 'A'
        
        return val[ord(a[i]) - ord(AA)] < val[ord(b[i]) - ord(AA)]
    
    if len(a) == len(b):
        return False
    
    return True

if __name__ == "__main__":
    alphabet = input().strip()
    val = [0] * 26
    for i, curr in enumerate(alphabet):
        val[ord(curr) - ord('a')] = i

    n = int(input().strip())
    words = []
    for _ in range(n):
        words.append(input().strip())
    

    for i in range(n):
        for j in range(i + 1, n):
            if not compare(words[i], words[j]):
                words[i], words[j] = words[j], words[i]

    for word in words:
        print(word)
